src/adf-to-markdown.ts

```diff
@@ -0,0 +1,585 @@
+/**
+ * ADF to Obsidian Markdown Converter
+ * 
+ * Converts Atlassian Document Format (ADF) to Obsidian-compatible Markdown.
+ * Based on:
+ * - ADF spec: https://developer.atlassian.com/cloud/jira/platform/apis/document/structure/
+ * - Obsidian Markdown: https://help.obsidian.md/
+ */
+
+import type {
+	ADFDocument,
+	ADFMark,
+	ADFNode,
+	BlockquoteNode,
+	BulletListNode,
+	CodeBlockNode,
+	DateNode,
+	DecisionItemNode,
+	DecisionListNode,
+	EmojiNode,
+	ExpandNode,
+	HeadingNode,
+	InlineCardNode,
+	ListItemNode,
+	MediaNode,
+	MediaSingleNode,
+	MentionNode,
+	OrderedListNode,
+	PanelNode,
+	StatusNode,
+	TableCellNode,
+	TableHeaderNode,
+	TableNode,
+	TableRowNode,
+	TaskItemNode,
+	TaskListNode,
+	TextNode,
+} from './adf-types';
+
+/**
+ * Options for the ADF to Markdown conversion.
+ */
+export interface ADFToMarkdownOptions {
+	/**
+	 * Whether to include a comment for unsupported features.
+	 * Default: true
+	 */
+	includeUnsupportedComments?: boolean;
+
+	/**
+	 * Custom handler for media URLs (e.g., to resolve Confluence attachment URLs).
+	 * If not provided, uses the raw URL or a placeholder.
+	 */
+	mediaUrlResolver?: (mediaNode: MediaNode) => string;
+
+	/**
+	 * Custom handler for mention resolution (e.g., to convert user IDs to names).
+	 * If not provided, uses the text attribute or the ID.
+	 */
+	mentionResolver?: (mentionNode: MentionNode) => string;
+}
+
+/**
+ * Converts Atlassian Document Format (ADF) to Obsidian-compatible Markdown.
+ */
+export class ADFToMarkdownConverter {
+	private options: Required<ADFToMarkdownOptions>;
+
+	constructor(options: ADFToMarkdownOptions = {}) {
+		this.options = {
+			includeUnsupportedComments: options.includeUnsupportedComments ?? true,
+			mediaUrlResolver: options.mediaUrlResolver ?? this.defaultMediaUrlResolver.bind(this),
+			mentionResolver: options.mentionResolver ?? this.defaultMentionResolver.bind(this),
+		};
+	}
+
+	/**
+	 * Convert an ADF document to Obsidian Markdown.
+	 * @param adf - The ADF document to convert
+	 * @returns The converted Markdown string
+	 */
+	convert(adf: ADFDocument): string {
+		if (!adf || adf.type !== 'doc' || !Array.isArray(adf.content)) {
+			return '';
+		}
+
+		const result = this.convertNodes(adf.content);
+		return result.trim();
+	}
+
+	/**
+	 * Convert an array of ADF nodes to Markdown.
+	 */
+	private convertNodes(nodes: ADFNode[], context: ConversionContext = {}): string {
+		return nodes.map(node => this.convertNode(node, context)).join('');
+	}
+
+	/**
+	 * Convert a single ADF node to Markdown.
+	 */
+	private convertNode(node: ADFNode, context: ConversionContext = {}): string {
+		switch (node.type) {
```

Comment:
let's implement this with a dictionary from node.Type -> Converter interface. Make the singular interface as a superset of all of these function calls, and let individual class implementations assume conversion from node to concrete object. Make all of the classes and put them into a directory